File: frontend/e2e/perf/async-actions.perf.spec.ts
import { test, expect } from "../fixtures/auth";
import {
  measureInteraction,
  expectAcknowledgementWithin,
  expectCompletionWithin,
  expectStableInteraction,
} from "./helpers/perfAssertions";
import { ACK_BUDGET, COMPLETION_BUDGET } from "./helpers/flowBudgets";

test.describe("Async action responsiveness", () => {
  test("run signup completes within budget", async ({ page }) => {
    await page.goto("/runs?run=run-public-empty-deadmines");
    const signupRegion = page
      .getByTestId("run-card")
      .filter({ hasText: "Public dungeon warmup" })
      .getByRole("region", { name: "Your Signup for Public dungeon warmup" });

    // Wait for characters to load (spinner disappears, attendance buttons visible)
    await signupRegion.getByRole("button", { name: "Late" }).waitFor({ state: "visible" });

    const cancelButton = signupRegion.getByRole("button", { name: "Cancel" });

    const result = await measureInteraction(
      page,
      () => signupRegion.getByRole("button", { name: "Late" }).click(),
      {
        ackMarker: cancelButton,
        completionMarker: cancelButton,
      },
    );

    // Verify the signup actually completed
    await expect(signupRegion.getByRole("button", { name: "Late" })).toHaveAttribute("aria-pressed", "true");

    expectAcknowledgementWithin(result, COMPLETION_BUDGET.NETWORK);
    expectCompletionWithin(result, COMPLETION_BUDGET.NETWORK);
    expectStableInteraction(result);
  });

  test("cancel signup completes within budget", async ({ page }) => {
    await page.goto("/runs?run=run-public-existing-signup-onyxia25");
    const signupRegion = page
      .getByTestId("run-card")
      .filter({ hasText: "Dragon reset clear" })
      .getByRole("region", { name: "Your Signup for Dragon reset clear" });

    // Wait for existing signup to render
    await expect(signupRegion.getByText("Aelrin")).toBeVisible();

    // Enter cancel confirmation
    await signupRegion.getByRole("button", { name: "Cancel" }).click();
    const cancelDialog = page.getByRole("dialog", { name: "Cancel signup?" });
    await cancelDialog.waitFor({ state: "visible" });

    // Completion: Character select reappears after cancel succeeds
    const characterSelect = signupRegion.getByLabel("Character");

    const result = await measureInteraction(
      page,
      () => cancelDialog.getByRole("button", { name: "Cancel signup" }).click(),
      {
        ackMarker: characterSelect,
        completionMarker: characterSelect,
      },
    );

    // Verify cancel completed — Cancel button should be gone
    await expect(signupRegion.getByRole("button", { name: "Cancel" })).toHaveCount(0);

    expectAcknowledgementWithin(result, COMPLETION_BUDGET.NETWORK);
    expectCompletionWithin(result, COMPLETION_BUDGET.NETWORK);
    expectStableInteraction(result);
  });
});

File: tests/Lfm.E2E/Perf/AsyncActionsPerfSpec.cs
using FluentAssertions;
using Lfm.E2E.Fixtures;
using Lfm.E2E.Helpers;
using Microsoft.Playwright;
using Xunit;

namespace Lfm.E2E.Perf;

/// <summary>
/// Perf port of frontend/e2e/perf/async-actions.perf.spec.ts — async action responsiveness.
/// Tagged [Trait("Category", "Perf")] so CI can exclude with --filter "Category!=Perf".
/// Collection: default (full seed data, authenticated context).
/// </summary>
[Collection("default")]
public class AsyncActionsPerfSpec(DefaultSeedFixture fixture) : IAsyncLifetime
{
    private IBrowserContext _context = null!;
    private IPage _page = null!;

    public async Task InitializeAsync()
    {
        _context = await AuthHelper.CreateAuthenticatedContextAsync(
            fixture.Browser, fixture.ApiBaseUrl, fixture.AppBaseUrl);
        _page = await _context.NewPageAsync();
    }

    public async Task DisposeAsync()
    {
        await _context.CloseAsync();
    }

    [Fact]
    [Trait("Category", "Perf")]
    public async Task Run_signup_completes_within_budget()
    {
        await _page.GotoAsync(fixture.AppBaseUrl + "/runs?run=run-public-empty-deadmines");
        var signupRegion = _page
            .GetByTestId("run-card")
            .Filter(new() { HasText = "Public dungeon warmup" })
            .GetByRole(AriaRole.Region, new() { Name = "Your Signup for Public dungeon warmup" });

        // Wait for characters to load (spinner disappears, attendance buttons visible)
        await signupRegion.GetByRole(AriaRole.Button, new() { Name = "Late" })
            .WaitForAsync(new() { State = WaitForSelectorState.Visible });

        var cancelButton = signupRegion.GetByRole(AriaRole.Button, new() { Name = "Cancel" });

        var result = await PerfHelper.MeasureInteractionAsync(
            _page,
            () => signupRegion.GetByRole(AriaRole.Button, new() { Name = "Late" }).ClickAsync(),
            ackMarker: cancelButton,
            completionMarker: cancelButton);

        // Verify the signup actually completed
        await Expect(signupRegion.GetByRole(AriaRole.Button, new() { Name = "Late" }))
            .ToHaveAttributeAsync("aria-pressed", "true");

        PerfHelper.AssertAcknowledgementWithin(result, CompletionBudget.Network);
        PerfHelper.AssertCompletionWithin(result, CompletionBudget.Network);
        PerfHelper.AssertStableInteraction(result);
    }

    [Fact]
    [Trait("Category", "Perf")]
    public async Task Cancel_signup_completes_within_budget()
    {
        await _page.GotoAsync(fixture.AppBaseUrl + "/runs?run=run-public-existing-signup-onyxia25");
        var signupRegion = _page
            .GetByTestId("run-card")
            .Filter(new() { HasText = "Dragon reset clear" })
            .GetByRole(AriaRole.Region, new() { Name = "Your Signup for Dragon reset clear" });

        // Wait for existing signup to render
        await Expect(signupRegion.GetByText("Aelrin")).ToBeVisibleAsync();

        // Enter cancel confirmation
        await signupRegion.GetByRole(AriaRole.Button, new() { Name = "Cancel" }).ClickAsync();
        var cancelDialog = _page.GetByRole(AriaRole.Dialog, new() { Name = "Cancel signup?" });
        await cancelDialog.WaitForAsync(new() { State = WaitForSelectorState.Visible });

        // Completion: character select reappears after cancel succeeds
        var characterSelect = signupRegion.GetByLabel("Character");

        var result = await PerfHelper.MeasureInteractionAsync(
            _page,
            () => cancelDialog.GetByRole(AriaRole.Button, new() { Name = "Cancel signup" }).ClickAsync(),
            ackMarker: characterSelect,
            completionMarker: characterSelect);

        // Verify cancel completed — Cancel button should be gone
        await Expect(signupRegion.GetByRole(AriaRole.Button, new() { Name = "Cancel" }))
            .ToHaveCountAsync(0);

        PerfHelper.AssertAcknowledgementWithin(result, CompletionBudget.Network);
        PerfHelper.AssertCompletionWithin(result, CompletionBudget.Network);
        PerfHelper.AssertStableInteraction(result);
    }

    private static ILocatorAssertions Expect(ILocator locator) =>
        Microsoft.Playwright.Assertions.Expect(locator);
}
